zulipterminal/ui_tools/boxes.py

```diff
@@ -808,6 +812,32 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
         elif is_command_key("MARKDOWN_HELP", key):
             self.view.controller.show_markdown_help()
             return key
+        elif is_command_key("OPEN_EXTERNAL_EDITOR", key):
+            editor = os.environ.get("ZULIP_EDITOR_COMMAND", os.environ.get("EDITOR"))
+            if editor is None:
+                self.view.controller.report_error(
+                    "Configure $EDITOR or $ZULIP_EDITOR_COMMAND shell environment."
+                )
+                return key
+            editor_splits = editor.split(" ")
```

Comment:
Splitting on all spaces seems problematic, especially on platforms with crazy paths like Windows (`"C:\Program Files (x86)\Notepad++\Notepad++.exe"`) and macOS (`"/Applications/Sublime Text.app"`).

On POSIX, the right behavior would be `editor_splits = shlex.split(editor)` ([documentation](https://docs.python.org/3/library/shlex.html#shlex.split)). I’m not sure if Windows users would find it intuitive that this requires them to quote `"C:\Windows\notepad.exe"`, though (python/cpython#44990). `shlex.split(editor, posix=False)` doesn’t really help.

```pycon
>>> # unintuitive?
>>> shlex.split(r'"C:\Windows\notepad.exe"')
['C:\\Windows\\notepad.exe']
>>> # too much splitting
>>> shlex.split(r'C:\Program Files (x86)\Notepad++\Notepad++.exe', posix=False)
['C:\\Program', 'Files', '(x86)\\Notepad++\\Notepad++.exe']
>>> # extra quotes left behind
>>> shlex.split(r'"C:\Program Files (x86)\Notepad++\Notepad++.exe"', posix=False)
['"C:\\Program Files (x86)\\Notepad++\\Notepad++.exe"']
```

So I’m not sure what to suggest.

Somehow Python on Windows manages to guess the intended parsing of `subprocess.check_call(r'C:\Program Files (x86)\notepad++\notepad++.exe test.txt')`, but I don’t understand what rules it uses to do that, and I can’t imagine how they could be secure…

Reply:
Maybe the [`oslex` library](https://pypi.org/project/oslex/) does the right thing? You’d think there’d be a library with more than one GitHub star for this.

Reply:
I took a look at oslex, even with one star the code looks good. Should we add this dependency or copy the 50 like lines of code in some utils (license is MIT).

Reply:
But even if we include the os detection in zulip we would still need [mslex](https://github.com/smoofra/mslex) dependency.

Reply:
Based on the examples above, it appears that `shlex.split` fails to behave as one might expect for *any* path with an unescaped space in eg. `Sublime Text.app some_arg` gives two arguments to `Sublime`. So in any case we'd need to note that users must be sure to escape spaces in file names, in these environment variables, if we used some form of `shlex.split`. eg. `export EDITOR=Sublime\ Text.app some_arg`
- Is that already necessary if one wants to use Sublime as the default $EDITOR on Mac?
- Surely other tools which use $EDITOR have similar parsing challenges?

The `oslex` and `mslex` docs lead to some brain-melting reading on windows path parsing! A potential simplification is that we don't support Windows natively (#357), only via WSL, so we're always expecting unix-like Python behavior (except in a few cases with temporary files). By default, executables in the windows path can be run from WSL, and I expect would also be executable via somewhere in eg. `/mnt/c/`, which would suggest we could rely on unix-like quoting rules in those cases. That behavior suggests `C:\...`-style paths inside WSL are not usable? There are some nested commands which could be more complex, maybe eg. `CMD.exe`, but indications are that MS aims to integrate win32/linux binaries into whatever environment you're in, so we may be able to just avoid the windows parsing pain when in WSL.

One wrinkle is that environment variables can be shared from windows to WSL, and don't always do automatic path translation, but we can document that.

So:
- `shlex.split` seems better than basic `split`, but still needs special care
- It would be good to confirm, but I don't believe we need to handle windows-style paths via `mslex`, since our supported platforms couldn't access them in any case
- Can we just document expected parsing challenges from shlex and WSL? Would that be insecure?
- Can we find inspiration from other applications using `$EDITOR`? Are they simpler, or also potentially insecure? Are they less cross-platform?

Other than trusting users follow the documentation, we could explicitly expect them to enter different parts of the command, ie. require variables for the command and (separate) arguments. That could be messy with multiple environment variables, particularly if trying to reuse `$EDITOR`, though simpler with configuration options.